### PR TITLE
added specialisation for parallelPlusDense in kmeans

### PR DESCRIPTION
**Pull request by Alexander-Makaryev**
Implementation of https://github.com/IntelPython/daal4py/issues/77

### CODE REVIEW

File: src/dist_kmeans_init.h [hunk 1]
```diff
@@ -197,6 +201,117 @@ namespace dist_custom {
...
+
+                const int S34TAG = 3003;
+                // The input for s4 will be stored in s4In
+                daal::data_management::NumericTablePtr s4In;
+                // run step3 on root and send results to non-roots
+                if(rank == 0) {
+                    step3Output = algo.run_step3Master(s3In);
+                    s4In = step3Output->get(daal::algorithms::kmeans::init::outputOfStep3ForStep4, 0);
+                    for(int i=1; i<nRanks; i++) {
+                        MPI4DAAL::send(step3Output->get(daal::algorithms::kmeans::init::outputOfStep3ForStep4, i), i, S34TAG); // it can be NULL
```
Comment (fschlimb) [inline src/dist_kmeans_init.h]: Why don't you follow what is done in plusplus method?
What does this change except that it makes it less scalable?

File: src/mpi4daal.h [hunk 2]
```diff
@@ -123,11 +129,12 @@ struct MPI4DAAL
...
-            all.resize(nRanks);
             for(int i=0; i<nRanks; ++i) {
-                // FIXME: This is super-inefficient, we need to write our own DatArchive to avoid extra copy
-                daal::data_management::OutputDataArchive out_arch(reinterpret_cast<daal::byte*>(buff+offsets[i]), sizes[i]);
-                all[i] = daal::services::staticPointerCast<typename T::ElementType>(out_arch.getAsSharedPtr());
+                if (sizes[i] > 0) {
+                    // FIXME: This is super-inefficient, we need to write our own DatArchive to avoid extra copy
+                    daal::data_management::OutputDataArchive out_arch(reinterpret_cast<daal::byte*>(buff+offsets[i]), sizes[i]);
+                    all.push_back(daal::services::staticPointerCast<typename T::ElementType>(out_arch.getAsSharedPtr()));
+                }
```
Comment (fschlimb) [inline src/mpi4daal.h]: Again, why do you make a change which makes things less efficient?

File: src/dist_kmeans_init.h [hunk 3]
```diff
@@ -20,6 +20,10 @@
 #include "dist_custom.h"
 #include "map_reduce_star.h"
 
+using namespace daal;
+using namespace daal::algorithms;
+using namespace daal::services;
```
Comment (fschlimb) [inline src/dist_kmeans_init.h]: Please do not use 'using'

**Comment by fschlimb** [comment]
@Alexander-Makaryev pls rebase onto master and force push, we need the CIs green.

**Comment by Alexander-Makaryev** [comment]
@oleksandr-pavlyk please, look at fails

**Comment by oleksandr-pavlyk** [comment]
Two failing tests in patches run, which are not failing in the unpatched one are:

```
FAILED miniconda/envs/bld/lib/python3.6/site-packages/sklearn/tests/test_discriminant_analysis.py::test_lda_dimension_warning[5-3]
FAILED miniconda/envs/bld/lib/python3.6/site-packages/sklearn/tests/test_discriminant_analysis.py::test_lda_dimension_warning[3-5]
```

They are sporadically failing in the unpatched run as well, so I have re-triggered the workflow.

**Comment by oleksandr-pavlyk** [comment]
The warning, that triggers the failure comes from true_divide:

```
>                                       ', '.join(str(warning) for warning in w)))
E               AssertionError: Got warnings when calling fit: [{message : RuntimeWarning('invalid value encountered in true_divide',), category : 'RuntimeWarning', filename : '/home/circleci/miniconda/envs/bld/lib/python3.6/site-packages/sklearn/discriminant_analysis.py', lineno : 92, line : None}]
```

**Comment by oleksandr-pavlyk** [comment]
The script I used to check that it works fine locally:

```python
import sklearn, numpy as np
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
(n_samples, n_features, n_classes) = (10, 5, 3)
rng = np.random.RandomState(0)
X = rng.randn(n_samples, n_features)
y = np.tile(range(n_classes), n_samples // n_classes + 1)[:n_samples]
y
max_components = min(n_features, n_classes - 1)
LinearDiscriminantAnalysis(n_components=2).fit(X, y) # no warning expected
```

**Comment by oleksandr-pavlyk** [comment]
For a non-zero probability of success in independent trials, with probability 1 the number of trials till green build is finite :)

**Comment by fschlimb** [comment]
So is there an issue in sklearn's test-suite?

**Comment by oleksandr-pavlyk** [comment]
Perhaps, but to have a better understanding and to suggest an improvement I need to be able to reproduce this problem to triage the root-cause.

13-15 failing tests in the released scikit-learn are all likely caused by a small set of common underlying problems which would be nice to identify.

**Comment by oleksandr-pavlyk** [comment]
I think that deselecting these 13-15 tests is not an option. I will try playing with CircleCI's `config.yml` to see if switching entirely to conda-forge channel would resolve this issue.

**Comment by Alexander-Makaryev** [comment]
@fschlimb I rebased it on latest master, for changes with parallel++ I will create new branch, CI is green, please review it one more time

**Pull request merged**